urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/MuxServer.java

```diff
@@ -0,0 +1,566 @@
+			while (!shutdown) {
+				try {
+					final Socket socket = listen.accept();
+					try {
+						context.registerSocket(socket);
+						final Integer pReadTimeout = inboundAddress.getOpts().getInteger(
+								Options.P_READ_TIMEOUT);
+						if (pReadTimeout != null) {
+							socket.setSoTimeout(pReadTimeout.intValue());
+						}
+						if (socket instanceof SSLSocket) {
+							((SSLSocket) socket).startHandshake();
+						}
+						Log.info(this.getClass().getSimpleName() + " new socket: " + socket + " "
+								+ SSLFactory.getSocketProtocol(socket) + " SendBufferSize="
+								+ socket.getSendBufferSize() + " ReceiveBufferSize="
+								+ socket.getReceiveBufferSize());
+						attender(socket);
+					} catch (Exception e) {
+						Log.error(this.getClass().getSimpleName() + " Exception: " + e.toString(), e);
+						context.closeSilent(socket);
+					}
+				} catch (SocketTimeoutException e) {
+					continue;
+				} catch (Exception e) {
+					if (!shutdown) {
+						Log.error(this.getClass().getSimpleName() + " " + e.toString(), e);
+					}
+					doSleep(1000);
+				}
+			}
+			close();
+			Log.info(this.getClass().getSimpleName() + " await end");
+			context.awaitShutdown(this);
+			Log.info(this.getClass().getSimpleName() + " end");
+		}
+
+		//
+		protected abstract void attender(Socket socket) throws IOException;
+	}
+
+	class MuxServerListenLocal extends MuxServerListen { // Local is MUX
+		MuxServerListenLocal(final InboundAddress inboundAddress) throws IOException {
+			super(inboundAddress);
+		}
+
+		@Override
+		protected synchronized void attender(final Socket socket) throws IOException {
+			Log.info(this.getClass().getSimpleName() + " attending socket: " + socket);
+			if ((local == null) || local.isClosed()) {
+				try {
+					socket.setSoTimeout(Constants.MUX_READ_TIMEOUT);
+				} catch (Exception ign) {
+				}
+				local = new MuxServerLocal(socket, inboundAddress);
+				local.setRouter(router);
+				context.addShutdownable(local);
+				listenRemote();
+				context.submitTask(local, "MuxInLeft[" + left + "|" + IOHelper.socketRemoteToString(socket)
+						+ "]", ClientId.newId());
+			} else {
+				// Only one concurrent client, close the new connection
+				Log.error(this.getClass().getSimpleName()
+						+ " This listener already connected, closing socket: " + socket);
+				sendNOP(); // FIXME: Try to check
+				doSleep(1000);
+				context.closeSilent(socket);
+			}
+		}
+	}
+
+	class MuxServerListenRemote extends MuxServerListen { // Remote is RAW
+		MuxServerListenRemote(final InboundAddress inboundAddress) throws IOException {
+			super(inboundAddress);
+		}
+
+		@Override
+		protected synchronized void attender(final Socket socket) throws IOException {
+			Log.info(this.getClass().getSimpleName() + " attending socket: " + socket);
+			final MuxServerRemote remote = new MuxServerRemote(socket, inboundAddress);
+			remote.setRouter(router);
+			mapRemotes.put(remote.getId(), remote);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **282**. At this line the following field is not protected by synchronization: `mapRemotes`. Your code typically protects this field with synchronization on object: `mapRemotes`.

**Fix**  
 Consider protecting the above field with synchronization on object: `mapRemotes`. For example, like at lines: **117**, **86**, **410**, and **407**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **86**, **117**, **407**, and **410**.